chainindex/read.go

```diff
@@ -0,0 +1,148 @@
+		return cid.Undef, xerrors.Errorf("failed to cast message CID: %w", err)
+	}
+
+	return msgCid, nil
+}
+
+func (si *SqliteIndexer) queryMsgCidFromEthHash(ctx context.Context, txHash ethtypes.EthHash, msgCidBytes *[]byte) error {
+	return si.getMsgCidFromEthHashStmt.QueryRowContext(ctx, txHash.String()).Scan(msgCidBytes)
+}
+
+func (si *SqliteIndexer) GetMsgInfo(ctx context.Context, messageCid cid.Cid) (*MsgInfo, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		si.closeLk.RUnlock()
+		return nil, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var tipsetKeyCidBytes []byte
+	var height int64
+
+	if err := si.readWithHeadIndexWait(ctx, func() error {
+		return si.queryMsgInfo(ctx, messageCid, &tipsetKeyCidBytes, &height)
+	}); err != nil {
+		return nil, err
+	}
+
+	tipsetKey, err := cid.Cast(tipsetKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+
+	return &MsgInfo{
+		Message: messageCid,
+		TipSet:  tipsetKey,
+		Epoch:   abi.ChainEpoch(height),
+	}, nil
+}
+
+// This function attempts to read data using the provided readFunc.
+// If the initial read returns no rows, it waits for the head to be indexed
+// and tries again. This ensures that the most up-to-date data is checked.
+// If no data is found after the second attempt, it returns ErrNotFound.
+func (si *SqliteIndexer) readWithHeadIndexWait(ctx context.Context, readFunc func() error) error {
+	err := readFunc()
+	if err == sql.ErrNoRows {
+		// not found, but may be in latest head, so wait for it and check again
+		if err := si.waitTillHeadIndexed(ctx); err != nil {
+			return xerrors.Errorf("failed while waiting for head to be indexed: %w", err)
+		}
+		err = readFunc()
+	}
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return ErrNotFound
+		}
+		return xerrors.Errorf("failed to get message info: %w", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) queryMsgInfo(ctx context.Context, messageCid cid.Cid, tipsetKeyCidBytes *[]byte, height *int64) error {
+	return si.getNonRevertedMsgInfoStmt.QueryRowContext(ctx, messageCid.Bytes()).Scan(tipsetKeyCidBytes, height)
+}
+
+func (si *SqliteIndexer) waitTillHeadIndexed(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, headIndexedWaitTimeout)
+	defer cancel()
+
+	head := si.cs.GetHeaviestTipSet()
+	headTsKeyCidBytes, err := toTipsetKeyCidBytes(head)
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// is it already indexed?
+	if exists, err := si.isTipsetIndexed(ctx, headTsKeyCidBytes); err != nil {
+		return xerrors.Errorf("failed to check if tipset exists: %w", err)
+	} else if exists {
+		return nil
+	}
+
+	// wait till it is indexed
+	subCh, unsubFn := si.subscribeUpdates()
+	defer unsubFn()
+
+	for ctx.Err() == nil {
+		exists, err := si.isTipsetIndexed(ctx, headTsKeyCidBytes)
+		if err != nil {
+			return xerrors.Errorf("failed to check if tipset exists: %w", err)
+		} else if exists {
+			return nil
+		}
+
```

Comment:
Two queries in quick succession, I get why you're doing this but given that a query involves a disk round-trip or five, might it be more efficient to just skip the first check and go straight into the subscribe + check loop? the biggest costs in a subscription are a mutex lock and map modification, so it's not very onerous.

Also, could we keep a `head` around on `SqliteIndexer` and check with that first? It may not be the right head, but it'd be an efficient check for the common case because you're comparing HeaviestTipSet here with what's in the db, but we could update our own `head` after successful indexing and we'd have a likely match when you enter this function. Could be gated by its own rwlock too.

Reply:
@rvagg I have removed the first read from DB and have gone straight to the "subscribe + check loop" as you suggest here. 

Tracking the `head` in the `SqliteIndexer ` is not something I want to introduce right now given the additional complexity of getting that tracking right and race free.

Reply:
ok, but this is in a fairly hot path though, `GetCidFromHash` and `GetMsgInfo` both block on making this query before making their own, the common path will be "our local `head` is the same as the reported chainstore head, no need to wait"

Reply:
@rvagg 

Not sure what you mean here. Both `GetCidFromHash ` and `GetMsgInfo ` will first try and read directly from the DB before doing the `waitTillHeadIndexed` thing. See:

```func (si *SqliteIndexer) readWithHeadIndexWait(ctx context.Context, readFunc func() error) error {
	err := readFunc()
	if err == sql.ErrNoRows {
		// not found, but may be in latest head, so wait for it and check again
		if err := si.waitTillHeadIndexed(ctx); err != nil {
			return xerrors.Errorf("failed while waiting for head to be indexed: %w", err)
		}
		err = readFunc()
	}

	if err != nil {
		if err == sql.ErrNoRows {
			return ErrNotFound
		}
		return xerrors.Errorf("failed to read data from index: %w", err)
	}

	return nil
}
```

Reply:
oh, right, goodo then!